### PR TITLE
Add delete support to upsert tables

### PR DESCRIPTION
**Pull request by navina**
This PR adds support for deleting records in an upsert table #10452  . The design details can be found [here](https://docs.google.com/document/d/19s1AHCRjmqeVa0z_djYBTAt_meg6qWcUxwRO3y-G9O4/edit#). 
The solution: 
* allows specifying the delete column name in the upsert config
* the delete operation updates the validDocIds bitmap and a dedicated bitmap to mark the deleted docId

Adding more tests to the PR is in progress. 

Labels: `feature` 

cc: @Jackie-Jiang @mayankshriv @npawar @yupeng9

### CODE REVIEW

**Comment by navina** [comment]
cc: @KKcorps

**Comment by KKcorps** [comment]
Should the `queryableDocIdSnapshot` needs to be persisted as well on shutdown for this to work?

Also, `queryableDocId` sounds confusing because `validDocId` is also used for querying. I think we should rename it to `activeRowIds` or `undeletedRowIds`

**Comment by Jackie-Jiang** [comment]
> Should the `queryableDocIdSnapshot` needs to be persisted as well on shutdown for this to work?

We don't need to persist it because we still need to load everything in the `validDocIds` (including the deleted docs) to refresh the timestamp for the deleted docs

> Also, `queryableDocId` sounds confusing because `validDocId` is also used for querying. I think we should rename it to `activeRowIds` or `undeletedRowIds`

If delete is enabled (by setting the `deleteRecordColumn`), we will only look at `queryableDocIds` during query time. We can add some javadoc so that it is more clear

File: pinot-integration-test-base/src/test/java/org/apache/pinot/integration/tests/BaseClusterIntegrationTest.java [hunk 1]
```diff
@@ -409,7 +416,62 @@ protected TableConfig createUpsertTableConfig(File sampleAvroFile, String primar
         .setRoutingConfig(new RoutingConfig(null, null, RoutingConfig.STRICT_REPLICA_GROUP_INSTANCE_SELECTOR_TYPE))
         .setSegmentPartitionConfig(new SegmentPartitionConfig(columnPartitionConfigMap))
         .setReplicaGroupStrategyConfig(new ReplicaGroupStrategyConfig(primaryKeyColumn, 1))
-        .setUpsertConfig(new UpsertConfig(UpsertConfig.Mode.FULL)).build();
+        .setUpsertConfig(upsertConfig).build();
+  }
+
+
```
Comment (Jackie-Jiang) [inline pinot-integration-test-base/src/test/java/org/apache/pinot/integration/tests/BaseClusterIntegrationTest.java:422]: (nit) extra empty line

File: pinot-integration-test-base/src/test/java/org/apache/pinot/integration/tests/BaseClusterIntegrationTest.java [hunk 2]
```diff
@@ -409,7 +416,62 @@ protected TableConfig createUpsertTableConfig(File sampleAvroFile, String primar
...
+  protected Map<String, String> getCSVStreamConfigMap(@Nullable String delimiter, @Nullable String csvHeaderProperty) {
```
Comment (Jackie-Jiang) [inline pinot-integration-test-base/src/test/java/org/apache/pinot/integration/tests/BaseClusterIntegrationTest.java:423]: This is not really the stream config. Suggest merging this into `createCSVUpsertTableConfig()`

File: pinot-integration-test-base/src/test/java/org/apache/pinot/integration/tests/BaseClusterIntegrationTest.java [hunk 3]
```diff
@@ -409,7 +416,62 @@ protected TableConfig createUpsertTableConfig(File sampleAvroFile, String primar
...
+      streamConfigsMap.put(StreamConfigProperties.constructStreamProperty(streamType, "decoder.prop.header"),
+          csvHeaderProperty);
+    }
+    return streamConfigsMap;
+  }
+
+  /**
+   * Creates a new Upsert enabled table config.
+   */
+  protected TableConfig createCSVUpsertTableConfig(String tableName, @Nullable String schemaName,
```
Comment (Jackie-Jiang) [inline pinot-integration-test-base/src/test/java/org/apache/pinot/integration/tests/BaseClusterIntegrationTest.java:443]: Remove the redundant arguments and use `getTableName()`, `getSchemaName()` etc.

File: pinot-integration-test-base/src/test/java/org/apache/pinot/integration/tests/BaseClusterIntegrationTest.java [hunk 4]
```diff
@@ -498,33 +560,82 @@ protected void setUpQueryGenerator(List<File> avroFiles) {
...
+  protected void pushCsvIntoKafka(File csvFile, String kafkaTopic, @Nullable Integer partitionColumnIndex) {
+    String kafkaBroker = "localhost:" + getKafkaPort();
+    StreamDataProducer producer = null;
+    try {
+      producer =
+        StreamDataProvider.getStreamDataProducer(KafkaStarterUtils.KAFKA_PRODUCER_CLASS_NAME,
+            getDefaultKafkaProducerProperties(kafkaBroker));
+      ClusterIntegrationTestUtils.pushCsvIntoKafka(csvFile, kafkaTopic, partitionColumnIndex, injectTombstones(),
+          producer);
+    } catch (Exception e) {
```
Comment (Jackie-Jiang) [inline pinot-integration-test-base/src/test/java/org/apache/pinot/integration/tests/BaseClusterIntegrationTest.java:608]: Suggest throwing the exception out. Currently it will fail silently. Same for the other method

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/MultiStageEngineIntegrationTest.java [hunk 5]
```diff
@@ -82,7 +82,7 @@ public void testHardcodedQueriesMultiStage()
     super.testHardcodedQueriesMultiStage();
   }
 
-  @Test
+  @Test (enabled = false)
```
Comment (Jackie-Jiang) [inline pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/MultiStageEngineIntegrationTest.java:85]: Remove this

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/UpsertTableSegmentUploadIntegrationTest.java [hunk 6]
```diff
@@ -74,7 +74,7 @@ public void setUp()
     // Create and upload schema and table config
     Schema schema = createSchema();
     addSchema(schema);
-    TableConfig tableConfig = createUpsertTableConfig(avroFiles.get(0), PRIMARY_KEY_COL, getNumKafkaPartitions());
+    TableConfig tableConfig = createUpsertTableConfig(avroFiles.get(0), PRIMARY_KEY_COL, null, getNumKafkaPartitions());
```
Comment (Jackie-Jiang) [inline pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/UpsertTableSegmentUploadIntegrationTest.java:77]: Consider merging this into the `UpsertTableIntegrationTest`? Essentially we want to upload one segment and check if it works properly

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/table/TableConfig.java [hunk 7]
```diff
@@ -388,6 +388,12 @@ public List<String> getUpsertComparisonColumns() {
     return _upsertConfig == null ? null : _upsertConfig.getComparisonColumns();
   }
 
+  @JsonIgnore
+  @Nullable
+  public String getUpsertDeleteRecordColumn() {
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/config/table/TableConfig.java:393]: Let's make sure the naming is consistent (either use `deleteRecordColumn` or `deletedRecordColumn`). I personally prefer `deleteRecordColumn` because it is used to delete a record

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManager.java [hunk 8]
```diff
@@ -220,11 +255,19 @@ protected void doAddRecord(MutableSegment segment, RecordInfo recordInfo) {
   protected GenericRow doUpdateRecord(GenericRow record, RecordInfo recordInfo) {
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManager.java:255]: Nice job handling the previous record deleted.

When the new record is a delete record, we can skip reading the previous record and updating the new record. We do want to still handle the out-of-order case. We can consider adding a flag for `outOfOrderEvent` to replace the current `previousRecordDeleted` flag

File: pinot-integration-test-base/src/test/java/org/apache/pinot/integration/tests/BaseClusterIntegrationTest.java (hunk 2 above)
Comment (navina) [reply pinot-integration-test-base/src/test/java/org/apache/pinot/integration/tests/BaseClusterIntegrationTest.java:423]: Yeah. this is just the decoder properties. Renaming the method appropriately.

File: pinot-integration-test-base/src/test/java/org/apache/pinot/integration/tests/BaseClusterIntegrationTest.java (hunk 3 above)
Comment (navina) [reply pinot-integration-test-base/src/test/java/org/apache/pinot/integration/tests/BaseClusterIntegrationTest.java:443]: Using these overrides will require writing a separate integration test class for every test. I wanted to avoid that and allow creating more than 1 table for an integration test class.

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/UpsertTableSegmentUploadIntegrationTest.java (hunk 6 above)
Comment (navina) [reply pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/UpsertTableSegmentUploadIntegrationTest.java:77]: The dataset used in that class is not really "upsert" like data. It has the airline data set with some 100K rows. I just wanted a simple functional test that is easy to read and doesn't require a large dataset. I can move the tests into a single class, if that's what you want.

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/table/TableConfig.java (hunk 7 above)
Comment (navina) [reply pinot-spi/src/main/java/org/apache/pinot/spi/config/table/TableConfig.java:393]: huh.. even I prefer `deleteRecordColumn`. Will fix the ones that deviated from this name.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManager.java (hunk 8 above)
Comment (navina) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManager.java:255]: > When the new record is a delete record, we can skip reading the previous record and updating the new record. 

Ah, I think it should still merge with previous record because technically, you could view it as just another update to the delete column.

Comment (Jackie-Jiang) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManager.java:255]: I don't follow. When it gets a delete record, the old record is considered deleted, and all other column values shouldn't matter anymore.

File: pinot-integration-tests/src/test/resources/partial_upsert_table_test.schema [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    }
+  ],
+  "dateTimeFieldSpecs": [{
+    "name": "timestampInEpoch",
+    "dataType": "LONG",
+    "format" : "1:MILLISECONDS:EPOCH",
+    "granularity": "1:MILLISECONDS"
+  }],
+  "primaryKeyColumns": [ "playerId" ]
+}
```
Comment (Jackie-Jiang) [inline pinot-integration-tests/src/test/resources/partial_upsert_table_test.schema:35]: Empty line after

File: pinot-integration-tests/src/test/resources/upsert_table_test.schema [hunk 10]
```diff
@@ -1,33 +1,34 @@
...
-}
+  ],
+  "dateTimeFieldSpecs": [{
+    "name": "timestampInEpoch",
+    "dataType": "LONG",
+    "format" : "1:MILLISECONDS:EPOCH",
+    "granularity": "1:MILLISECONDS"
+  }],
+  "primaryKeyColumns": [ "playerId" ]
+}
```
Comment (Jackie-Jiang) [inline pinot-integration-tests/src/test/resources/upsert_table_test.schema:34]: Empty line after

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManager.java [hunk 11]
```diff
@@ -221,18 +256,30 @@ protected void doAddRecord(MutableSegment segment, RecordInfo recordInfo) {
...
+              // if delete is not enabled or previous record not marked as deleted
+              _reuse.clear();
+              previousRecordReference.set(currentSegment.getRecord(currentDocId, _reuse));
+            }
+          } else {
+            outOfOrder.set(true);
           }
           return recordLocation;
         });
+    GenericRow previousRecord = previousRecordReference.get();
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManager.java:276]: Move this into the if check (revert the change)

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManager.java [hunk 12]
```diff
@@ -221,18 +256,30 @@ protected void doAddRecord(MutableSegment segment, RecordInfo recordInfo) {
   protected GenericRow doUpdateRecord(GenericRow record, RecordInfo recordInfo) {
     assert _partialUpsertHandler != null;
     AtomicReference<GenericRow> previousRecordReference = new AtomicReference<>();
+    AtomicBoolean outOfOrder = new AtomicBoolean();
     RecordLocation currentRecordLocation = _primaryKeyToRecordLocationMap.computeIfPresent(
         HashUtils.hashPrimaryKey(recordInfo.getPrimaryKey(), _hashFunction), (pk, recordLocation) -> {
           if (recordInfo.getComparisonValue().compareTo(recordLocation.getComparisonValue()) >= 0) {
-            _reuse.clear();
-            previousRecordReference.set(recordLocation.getSegment().getRecord(recordLocation.getDocId(), _reuse));
+            IndexSegment currentSegment = recordLocation.getSegment();
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManager.java:263]: Add an if check of `if (!recordInfo.isDeleteRecord())`. No need to read previous record when current record is delete

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManager.java [hunk 13]
```diff
@@ -221,18 +256,30 @@ protected void doAddRecord(MutableSegment segment, RecordInfo recordInfo) {
...
           }
           return recordLocation;
         });
+    GenericRow previousRecord = previousRecordReference.get();
     if (currentRecordLocation != null) {
       // Existing primary key
-      GenericRow previousRecord = previousRecordReference.get();
-      if (previousRecord != null) {
+      if (!outOfOrder.get()) {
+        if (recordInfo.isDeleteRecord() || previousRecord == null) {
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapPartitionUpsertMetadataManager.java:280]: We can simplify this check to `if (previousRecord == null)` after applying the check when reading the previous record

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/TableConfigUtils.java [hunk 14]
```diff
@@ -570,20 +570,29 @@ && isRoutingStrategyAllowedForUpsert(tableConfig.getRoutingConfig()),
...
         }
       }
+
+      // Delete record column exist and is a BOOLEAN field
+      String deleteRecordColumn = upsertConfig.getDeleteRecordColumn();
+      if (deleteRecordColumn != null) {
+        FieldSpec fieldSpec = schema.getFieldSpecFor(deleteRecordColumn);
+        Preconditions.checkState(
+            fieldSpec != null && fieldSpec.isSingleValueField() && fieldSpec.getDataType() == DataType.BOOLEAN,
+            "The deleted record column must be a single-valued BOOLEAN column");
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/TableConfigUtils.java:594]:
```suggestion
            "The delete record column must be a single-valued BOOLEAN column");
```

File: pinot-segment-local/src/test/java/org/apache/pinot/segment/local/utils/TableConfigUtilsTest.java [hunk 15]
```diff
@@ -1521,6 +1522,43 @@ public void testValidateUpsertConfig() {
...
+    upsertConfig.setDeleteRecordColumn(incorrectTypeDelCol);
+    tableConfig = new TableConfigBuilder(TableType.REALTIME).setTableName(TABLE_NAME).setStreamConfigs(streamConfigs)
+        .setUpsertConfig(upsertConfig)
+        .setRoutingConfig(new RoutingConfig(null, null, RoutingConfig.STRICT_REPLICA_GROUP_INSTANCE_SELECTOR_TYPE))
+        .build();
+    try {
+      TableConfigUtils.validateUpsertAndDedupConfig(tableConfig, schema);
+      Assert.fail("Invalid delete column type (string) should have failed table creation");
+    } catch (IllegalStateException e) {
+      Assert.assertEquals(e.getMessage(), "The deleted record column must be a single-valued BOOLEAN column");
```
Comment (Jackie-Jiang) [inline pinot-segment-local/src/test/java/org/apache/pinot/segment/local/utils/TableConfigUtilsTest.java:1548]:
```suggestion
      Assert.assertEquals(e.getMessage(), "The delete record column must be a single-valued BOOLEAN column");
```

**Review by Jackie-Jiang — APPROVED** [review]
LGTM otherwise
